<!DOCTYPE html>
<html>
    
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <title>Une introduction à Python pour les Math au lycée</title>
        <meta name="description" content="An interactive introduction to Python for Maths.">
        <link rel="stylesheet" href="main.css">
    </head>
    <body>
        <table width = 100%>
        <tbody>
        <tr><td><a href="index.html">Accueil</a></td>
            <td><a href="fonction.html">Première partie <br/>Fonction</a></td>
            <td><a href="sialors.html">Deuxième Partie <br/>Structure conditionnelle</a></td>
            <td><a href="pour.html">Troisième Partie <br/>Boucle POUR</a></td>
            <td><a href="tantque.html">Quatrième Partie<br/>Boucle TANT QUE</a></td></tr>
        <tr><td colspan = "2">Pour aller plus loin</td>
            <td>Les assertions</td>
            <td><a href="flottants.html">Les nombres flottants</a></td>
            <td><a href="modularite.html">La modularité</a></td></tr>
        </tbody>
        </table>
        <h1>Les assertions</h1>
        <h4> (ou gestion des erreurs)</h4>
        <p>
        En Mathématiques, on a souvent des fonctions qui ont des valeurs interdites, comme par exemple la fonction <math><mtable> <mtr>f(x) = <mfrac><mi>1</mi> <mi>x+1</mi></mfrac> </mtr> </mtable> </math> qui a pour valeur interdite -1. Il n'est pas rare de trouver ce genre de code, écrit par des élèves consciencieux :
        <pre>
    def f(x) :
        if x == -1:
            print("Attention, -1 est une valeur interdite !!!")
        else :
            return 1 / (x + 1)
        </pre>
        Pour des utilisations simples, cela ne pose pas de problème.
        <pre>
    >>> f(3)
    0.25
    >>> f(0)
    1.0
    >>> f(-1)
    Attention, -1 est une valeur interdite !!!
        </pre>
        </p>
    
    
        <p>
        Le problème apparaît si on appelle cette fonction dans une autre. En effet, pour la valeur -1, la fonction ne renvoie rien, elle affiche juste un  message.<br/>
        Par exemple, si on porgramme une fonction qui calcule un taux d'accroissement :
        <pre>
    def tauxAccroissement(f,a,b):
        return (f(a) - f(b)) / (a - b) 
        </pre>
        Cette fonction <code>tauxAccroissement</code> dépend de ce que la fonction <code>f</code> va renvoyer. Donc si on demande le taux d'accroissement de la fonction f entre 1 et 3, pas de souci.
        <pre>
    >>> tauxAccroissement(f,1,3)
    -0.125  
        </pre>
        Par contre si on demande le taux d'accroissement entre -1 et 3, le message d'erreur n'est pas si clair car la fonction <code>f</code> ne renvoie rien, elle affiche simplement un message. Donc la fonction  <code>tauxAccroissement</code> est incapable d'effectuer le calcul demandé. En outre le message d'erreur n'est pas clair pour tout le monde (pour la petite histoire comme <code>f</code> ne renvoie rien, le message d'erreur nous signale que Python est incapable d'effectuer une soustration entre un Rien et un flottant (nombre décimal)).
        <pre><xmp>
    >>> tauxAccroissement(f,-1,3)
    Attention, -1 est une valeur interdite
    Traceback (most recent call last):
      File "<input>", line 1, in <module>
      File "<input>", line 13, in tauxAccroissement
    TypeError: unsupported operand type(s) for -: 'NoneType' and 'float'
        </xmp></pre>
        </p>
    
    <h4>Une solution, les assertions</h4>
    L'idée est alors de s'assurer que la fonction est bien utilisée dans les conditions de son utilisation avec le mot clef <code>assert</code>, suivi éventuellement d'une virgule et d'un message d'erreur entre guillemet.
    Dans notre exemple, nous allons nous assurer que la fonction est utilisée pour des valeurs de <code>x</code> différentes de -1.
    <pre>
    def f_v2(x) :
        assert x != -1, "Attention, -1 est une valeur interdite"
        return 1 / (x + 1)
    </pre>
    Que se passe-t-il alors avec notre fonction <code>tauxAccroissement</code> ?
    <pre><xmp>
    >>> tauxAccroissement(f_v2,-1,3)
    Traceback (most recent call last):
      File "<input>", line 1, in <module>
      File "<input>", line 13, in tauxAccroissement
      File "<input>", line 8, in f_v2
    AssertionError: Attention, -1 est une valeur interdite
    </xmp></pre>
    La fonction <code>tauxAccroissement</code> est dans l'incapacité d'effectuer sa tache, car elle est bloquée au niveau de l'assertion de la fonction <code>f_v2</code> et surtout nous retrouvons notre message d'erreur end dernière ligne du message d'erreur, on comprend tout de suite le problème.<br/>
    <br/>
    On peut tester ce programme sur basthon en ligne sur le lien <a target="_blank" href="https://console.basthon.fr/?script=eJyNkE1qwzAQhfeB3OHVm1pUoaTdBWLISYosj4rAlox-jI-UnsMX6yiuIYUGOjuNZr735nVkYOpZ4LTfgcsazDifcTj-NEqNwbpUV5eUyCXrneRvUEzIjjCpnnIAT1DobKJKrIvUR8IdJFDKweGIV9QzXnDkuf2uK_of09udBRUjhcQ2nooNif_qrtuPZZLK80Xr4C0LDEysjVSyFadfi7WplcCBQ2mFKBDFj1agUKqKRUbPqoldUIAmzsZ_BjUMJKHGsFwjaF6-dF79dspFdstTNFgqe8MY126voL2LvicYZcMtq6Zp_rTJMbzzHRutozzb5TrQypObHUYy7yYNNZFG_7yGWRqP8Rz_psAXfgNXYp8Q">suivant</a>.
    </body>
</html>
